pkg/controller/volume/persistentvolume/scheduler_binder.go

```diff
@@ -253,58 +254,190 @@ func (b *volumeBinder) AssumePodVolumes(assumedPod *v1.Pod, nodeName string) (al
 		newProvisionedPVCs = append(newProvisionedPVCs, claimClone)
 	}
 
-	if len(newProvisionedPVCs) != 0 {
-		bindingRequired = true
+	// Update cache with the assumed pvcs and pvs
+	if len(newBindings) > 0 {
+		b.podBindingCache.UpdateBindings(assumedPod, nodeName, newBindings)
+	}
+	if len(newProvisionedPVCs) > 0 {
 		b.podBindingCache.UpdateProvisionedPVCs(assumedPod, nodeName, newProvisionedPVCs)
 	}
 
 	return
 }
 
-// BindPodVolumes gets the cached bindings and PVCs to provision in podBindingCache
-// and makes the API update for those PVs/PVCs.
+// BindPodVolumes gets the cached bindings and PVCs to provision in podBindingCache,
+// makes the API update for those PVs/PVCs, and waits for the PVCs to be completely bound
+// by the PV controller.
 func (b *volumeBinder) BindPodVolumes(assumedPod *v1.Pod) error {
 	podName := getPodName(assumedPod)
 	glog.V(4).Infof("BindPodVolumes for pod %q", podName)
 
 	bindings := b.podBindingCache.GetBindings(assumedPod, assumedPod.Spec.NodeName)
 	claimsToProvision := b.podBindingCache.GetProvisionedPVCs(assumedPod, assumedPod.Spec.NodeName)
 
+	// Start API operations
+	err := b.bindAPIUpdate(assumedPod, bindings, claimsToProvision)
+	if err != nil {
+		return err
+	}
+
+	// TODO: what's a good timeout value
+	return wait.Poll(time.Second, 10*time.Minute, func() (bool, error) {
+		return b.checkBindings(podName, bindings, claimsToProvision)
+	})
+}
+
+func getPodName(pod *v1.Pod) string {
+	return pod.Namespace + "/" + pod.Name
+}
+
+func getPVCName(pvc *v1.PersistentVolumeClaim) string {
+	return pvc.Namespace + "/" + pvc.Name
+}
+
+// bindAPIUpdate gets the cached bindings and PVCs to provision in podBindingCache
+// and makes the API update for those PVs/PVCs.
+func (b *volumeBinder) bindAPIUpdate(assumedPod *v1.Pod, bindings []*bindingInfo, claimsToProvision []*v1.PersistentVolumeClaim) error {
+	podName := getPodName(assumedPod)
+
+	unprocessedBindings := 0
```

Comment:
Hmm, better to name it "processedXXX", i.e. how many bindings/provisions have been processed